### PR TITLE
feat: find canonical ancestor

### LINKED ISSUE #2502: Return latest valid hash on invalid payload
### Describe the feature

the blockchaintree currently returns the combine interfaces error,

since the CL expects specific responses (valid hash, validation error) based on the error we need a more ergonomic way of handling this

ideally custom errors

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mattsse**
Closes #2502

on invalid payload, this tries to find the latest valid payload, by walking up the chain, see `latest_valid_hash_for_invalid_payload` docs.

### <samp>🤖 Generated by Copilot at a0300e3</samp>

This pull request adds the functionality of finding and returning the latest valid hash for an invalid payload, as part of the `engine_executePayloadV1` RPC method. It modifies the `Engine`, `ShareableBlockchainTree`, `BlockchainTree`, and `Provider` structs and traits, as well as the `PayloadStatus` struct and the `engine` and `blockchain-tree` modules. It also updates the relevant imports and comments.

### CODE REVIEW

File: crates/blockchain-tree/src/shareable.rs [hunk 1]
```diff
@@ -95,6 +95,21 @@ impl<DB: Database, C: Consensus, EF: ExecutorFactory> BlockchainTreeViewer
...
+
+        // walk up the tree and check if the parent is canonical
+        loop {
+            if let Some(block) = tree.block_by_hash(parent) {
+                parent = block.parent_hash;
+            }
+            if tree.block_indices().is_block_hash_canonical(&parent) {
+                return Some(parent)
+            }
+        }
```
Comment (rkrasiuk) [inline crates/blockchain-tree/src/shareable.rs:110]: what if the parent hash is not known to the tree? seems like the loop would never terminate

File: crates/consensus/beacon/src/engine/mod.rs [hunk 2]
```diff
@@ -251,6 +252,33 @@ where
...
+        // check pre merge block error
+        if let Some(Error::Execution(ExecutorError::BlockPreMerge { .. })) = tree_error {
+            return Some(H256::zero())
+        }
+
+        if let latest_valid @ Some(_) = self.blockchain_tree.find_canonical_ancestor(parent_hash) {
+            return latest_valid
+        }
+
+        None
```
Comment (rkrasiuk) [inline crates/consensus/beacon/src/engine/mod.rs:279]:
```suggestion
        self.blockchain_tree.find_canonical_ancestor(parent_hash)
```

**Review by rkrasiuk — CHANGES_REQUESTED** [review]

File: crates/blockchain-tree/src/shareable.rs (hunk 1 above)
Comment (mattsse) [reply crates/blockchain-tree/src/shareable.rs:110]: you're right, forgot the break

Comment (Rjected) [reply crates/blockchain-tree/src/shareable.rs]: ah yeah was going to comment this

**Review by Rjected — APPROVED** [review]
LGTM now

**Review by rkrasiuk — APPROVED** [review]
lgtm

**Pull request merged**